### PR TITLE
Moves Hands to shared, some prediction

### PR DESCRIPTION
**Pull request by collinlunn**
Moves a majority of hand code to shared, makes held item sprites be handled by a visualizer, implements prediction for swapping hands and dropping items.

A lot of stuff is WIP still but I want to consolidate what I've done so far, since PR size is piling up.

Todo:
- [ ] Locate the rest of the bugs
 - [x] Fix crash when entities inside hand delete themselves
 - [x] Make sure IsFirstTimePredicted is checked on sounds/visuals caused by current hands prediction

### CODE REVIEW

**Comment by ike709** [comment]
There's some sus changes in here, might want to test that this doesn't break HUD themes.

File: Content.Client/GameObjects/EntitySystems/HandsSystem.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            EntityManager.RaisePredictiveEvent(new DropMessage(coords));
+            return true;
+        }
+
+        protected override void HandleContainerModified(ContainerModifiedMessage args)
+        {
+            if (args.Container.Owner.TryGetComponent(out SharedHandsComponent? hands))
+                hands.HandsModified();
+        }
```
Comment (metalgearsloth) [inline Content.Client/GameObjects/EntitySystems/HandsSystem.cs:46]: We have directed versions of the eventbus for this now so you don't need to TryGet it, buckle uses it from memory.

**Comment by collinlunn** [comment]
Fixed what issues I could find with the changes. Hand code still needs work for further cleanup of obsolete methods, general code quality improvement, and to implement the rest of prediction.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
I'm not gonna be anal on ECS as this was done before it.

Personally I'd prefer that hands get refactored entirely in one go rather than piecemealed over PRs but this is fine for now once the reviews are done.

File: Content.Shared/GameObjects/EntitySystems/SharedHandsSystem.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        private void HandleDrop(DropMessage msg, EntitySessionEventArgs eventArgs)
+        {
+            var entity = eventArgs.SenderSession?.AttachedEntity;
+
+            if (entity == null || !entity.TryGetComponent(out SharedHandsComponent? hands))
+                return;
+
+            hands.TryDropActiveHand(msg.DropTarget);
+        }
```
[the remark this answers is not in the document]
Comment (collinlunn) [reply Content.Shared/GameObjects/EntitySystems/SharedHandsSystem.cs]: Client now specifies which hand it is dropping from, the prevent any desyncs from preventing the player form dropping from their intended hand.

File: Content.Shared/GameObjects/EntitySystems/SharedHandsSystem.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+        private void HandleSwapHands(SwapHandsMessage msg, EntitySessionEventArgs eventArgs)
+        {
+            var entity = eventArgs.SenderSession?.AttachedEntity;
+
+            if (entity == null || !entity.TryGetComponent(out SharedHandsComponent? hands))
+                return;
+
+            hands.SwapHands();
+        }
```
[the remark this answers is not in the document]
Comment (collinlunn) [reply Content.Shared/GameObjects/EntitySystems/SharedHandsSystem.cs]: Fixed the same way to the dropping event.

File: Content.Client/GameObjects/Components/Items/HeldItemsVisualizer.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                _layerKeys.Add(layerKey);
+                sprite.LayerMapReserveBlank(layerKey);
+
+                var layer = sprite.LayerMapGet(layerKey);
+                sprite.LayerSetVisible(layer, true);
+                sprite.LayerSetRSI(layer, rsi);
+                sprite.LayerSetState(layer, item.State);
+                sprite.LayerSetColor(layer, item.Color);
+
+            }
```
[the remark this answers is not in the document]
Comment (collinlunn) [reply Content.Client/GameObjects/Components/Items/HeldItemsVisualizer.cs]: Not sure how I'd go about that, I think I'll call that out of scope.

File: Content.Server/GameObjects/Components/GUI/HandsComponent.cs [hunk 5]
```diff
@@ -41,814 +35,350 @@ public class HandsComponent : SharedHandsComponent, IHandsComponent, IBodyPartAd
...
-        }
+            var outermostEntity = entity;
+            while (outermostEntity.TryGetContainer(out var container))
+                outermostEntity = container.Owner;
 
-        public bool CanPutInHand(ItemComponent item, bool mobCheck = true)
-        {
-            if (mobCheck && !ActionBlockerSystem.CanPickup(Owner))
-                return false;
+            var initialPosition = outermostEntity.Transform.Coordinates;
```
[the remark this answers is not in the document]
Comment (collinlunn) [reply Content.Server/GameObjects/Components/GUI/HandsComponent.cs]: I'm fairly sure the 0,0 lerp is fixed. WorldPosition would be cleaner that looping to find the outermost container, but I would need to to change the animation or make another coordinate conversion helper, which I am calling out of scope, putting a TODO on it.

File: Content.Client/GameObjects/Components/Items/HandsComponent.cs [hunk 6]
```diff
@@ -22,333 +22,199 @@ public class HandsComponent : SharedHandsComponent
...
+            Gui.Visible = true;
         }
 
-        public void AttackByInHand(string index)
+        private void HandlePlayerDetachedMsg()
         {
-            SendNetworkMessage(new ClientAttackByInHandMsg(index));
+            if (Gui != null)
+                Gui.Visible = false;
         }
```
[the remark this answers is not in the document]
Comment (collinlunn) [reply Content.Client/GameObjects/Components/Items/HandsComponent.cs]: I've cleaned the gui handling to hopefully be more clear that the subscriptions are handled right.

File: Content.Shared/GameObjects/EntitySystems/SharedHandsSystem.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            base.Initialize();
+
+            SubscribeLocalEvent<SharedHandsComponent, EntRemovedFromContainerMessage>(HandleContainerModified);
+            SubscribeLocalEvent<SharedHandsComponent, EntInsertedIntoContainerMessage>(HandleContainerModified);
+
+            SubscribeLocalEvent<SwapHandsMessage>(HandleSwapHands);
+            SubscribeNetworkEvent<SwapHandsMessage>(HandleSwapHands);
+
+            SubscribeLocalEvent<DropMessage>(HandleDrop);
+            SubscribeNetworkEvent<DropMessage>(HandleDrop);
```
[the remark this answers is not in the document]
Comment (collinlunn) [reply Content.Shared/GameObjects/EntitySystems/SharedHandsSystem.cs]: I'm not sure I fully understand what you are asking about.

Currently, the client issues a SwapHands event by `RaisePredictiveEvent` which raises both a local and networked version of the event. The local event triggers the predicted hand swap, and the networked event causes the server to run the hand swap and send back the updated state.

**Review by metalgearsloth — APPROVED** [review]

**Comment by metalgearsloth** [comment]
Fix conflict (Probably needs the namespace update too).

**Comment by collinlunn** [comment]
Went through and made sure everything is in the new appropriate folders and namespaces,

**Pull request merged**